src/provider/mod.rs

```diff
@@ -422,54 +582,170 @@ where
                 L1CommunicationError::Custom("Error occurred while estimating L1 base fees.")
             })?;
         let max_fee_per_gas = base_l1_fees_data.max_fee_per_gas + max_priority_fee_per_gas;
-        //
 
-        let l2_base_cost = bridge_hub_contract
-            .l2TransactionBaseCost(
-                U256::from(l2_chain_id),
-                U256::from(max_fee_per_gas),
-                estimate_gas_l1_to_l2,
-                gas_per_pubdata_limit,
-            )
-            .call()
+        let bridge_hub_contract_address = self
+            .get_bridgehub_contract()
             .await
             .map_err(|_| {
                 L1CommunicationError::Custom(
-                    "Error occurred while estimating L2 transaction base cost.",
+                    "Error occurred while fetching the bridge hub contract address.",
                 )
             })?
-            ._0;
+            .unwrap();
+        let bridge_hub_contract = Bridgehub::new(bridge_hub_contract_address, l1_provider);
 
-        let l1_value = l2_base_cost + amount;
+        if deposit_request.token != ETHER_L1_ADDRESS {
+            let (l1_bridge_address, l2_bridge_address) = self
+                .get_bridge_addresses_for_deposit(deposit_request, l2_chain_id, l1_provider)
+                .await?;
 
+            let erc20_contract = ERC20::new(deposit_request.token, l1_provider);
+            let token_data = encode_token_data_for_bridge(&erc20_contract)
+                .await
+                .map_err(|_| {
+                    L1CommunicationError::Custom("Error while encoding ERC20 token data.")
+                })?;
+            let l2_finalize_deposit_calldata = encode_finalize_deposit_calldata(
+                sender,
+                receiver,
+                deposit_request.token,
+                deposit_request.amount,
+                token_data,
+            );
+
+            let l2_tx_fee = self
+                .get_bridge_l2_tx_gas_data(
+                    &bridge_hub_contract,
+                    TransactionRequest::default()
+                        .with_from(apply_l1_to_l2_alias(l1_bridge_address))
+                        .with_to(l2_bridge_address)
+                        .with_gas_per_pubdata(deposit_request.gas_per_pubdata_limit)
+                        .with_input(l2_finalize_deposit_calldata),
+                    l2_chain_id,
+                    U256::from(max_fee_per_gas),
+                    deposit_request.gas_per_pubdata_limit,
+                )
+                .await?;
+
+            let token_allowance = erc20_contract
```

Comment:
I think `DepositRequest` should have a setting to either disable automatic approvals, smth like 

```
impl DepositRequest {
  /// Enable or disable automatic submission of ERC20 approval transactions
  /// if the allowance is not sufficient.
  fn with_auto_approval(auto_approval: bool) -> { .. }
}
```

Later we will probably want to make it possible to set custom amounts for approval, but it's not needed right now.

Reply:
Agreed, added the `auto_approval` option and the corresponding method to set it.